### PR TITLE
refactor react-overlays into individual modules

### PR DESCRIPTION
**Pull request by vitosamson**
also adds correct definitions for some modules whose definitions were incorrectly being imported from react-bootstrap. (react-bootstrap depends on react-overlays, not the other way around).

[This line](https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...vitosamson:react-overlays?expand=1#diff-f4cdead61a65aea2db80ddea3b56f4e5L9) was causing the tests in #18192 to fail. The modules re-exported in that line for some reason were declared as being pulled in from react-bootstrap, when really it's the other way around. Those modules were also being erroneously declared in the react-bootstrap typings, and so when I removed them, the tests for react-overlays failed.

This is blocking https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18192.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/react-bootstrap/react-overlays/tree/v0.6.12/src
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/react-overlays/lib/RootCloseWrapper.d.ts [hunk 1]
```diff
@@ -17,6 +23,4 @@ interface RootCloseWrapperProps {
 	event?: 'click' | 'mousedown';
 }
 
-declare class RootCloseWrapper extends React.Component<RootCloseWrapperProps> { }
-declare namespace RootCloseWrapper { } // module export workaround: https://github.com/Microsoft/TypeScript/issues/5073
-export = RootCloseWrapper;
+export default class RootCloseWrapper extends React.Component<RootCloseWrapperProps> { }
```
Comment (aaronbeall) [inline types/react-overlays/lib/RootCloseWrapper.d.ts]: Does this actually work correctly at runtime? When I originally wrote these a default export was not working.

**Review by aaronbeall — COMMENTED** [review]
Does react-bootstrap now inherit the type defs from react-bootstrap?

File: types/react-overlays/lib/RootCloseWrapper.d.ts (hunk 1 above)
Comment (vitosamson) [inline types/react-overlays/lib/RootCloseWrapper.d.ts]: Ah, it looks like it doesn't. Sorry I should have tested it more thoroughly. I'll fix this up tomorrow and will likely need to do the same for #18192. With that namespace hack, individual components will need to be imported with the `import * as ` syntax, right?

Is there really no other way to get interop between TS and babel's default exports without so much boilerplate?

**Comment by vitosamson** [comment]
@aaronbeall I've changed all the default exports to use the exported namespace hack.

re your question:

> Does react-bootstrap now inherit the type defs from react-bootstrap?

I assume you meant does react-bootstrap inherit type defs from react-overlays? If so, the answer is that react-bootstrap uses react-overlays internally, but doesn't expose any of its components or type defs.

**Comment by andy-ms** [comment]
Thanks!

**Comment by aaronbeall** [comment]
@vitosamson Sorry I didn't get back to this, I didn't get a chance to check these definitions against the actual react-overlays library but if they work then they look a great improvement. 

As for the question about inheritance, I thought that [react-bootstrap exposed some react-overlays components](https://react-bootstrap.github.io/components.html#overlays) like `<Modal>` and `<Overlay>` but now I see they wrap and extend them. I guess it's fine to just have duplicated definitions for what's shared...

**Pull request merged**